docs/Supplementary Views.md

```diff
@@ -0,0 +1,76 @@
+
+
+### How UtiliKit Helps
+
+UtiliKit aims to make this process as foolproof as possible. As always, there are 3 steps to correctly implementing the supplementary dequeue process- definition, registration, and dequeue. As an example, we'll walk through a case using a diffable data source and a compositional layout.
+
+1. Definition
+
+This occurs in the layout. In our case, when we are defining our `NSCollectionLayoutSection`. For example, a badge supplementary item on each of our cells may be defined something like this, where `UtiliKit` provides a small convenience here by providing some type-safety over the `String` implementation of `UIKit`:
+
+```swift
+extension UICollectionView.ElementKind {
+    static let badge = Self(rawValue: "badge-view")
+}
+
+func makeLayout() {
+  let badgeItem = NSCollectionLayoutSupplementaryItem(layoutSize: .init(widthDimension: .absolute(20), heightDimension: .absolute(20)),
+                                                      elementKind: .badge, containerAnchor: .init(edges: [.top, .leading]))
+
+  let itemSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1),
+                                        heightDimension: .fractionalHeight(1))
+  let item = NSCollectionLayoutItem(layoutSize: itemSize,
+                                    supplementaryItems: [badgeItem])
+
+  // rest of layout definition
+}
+```
+
+2. Registration
+
+The next step is then registering the class(es) or nib(s) that will fulfill this element kind in the collection view. In this example, there is only one, that we register in `viewDidLoad`, ensuring we pass the same `UICollectionView.ElementKind` in to the registration.
+
+`collectionView.register(BadgeSupplementaryView.self, forSupplementaryViewOfKind: .badge)`
+
+3. Dequeue
+
+The final step occurs when the collection view first requests a supplementary view for this element kind. In our case, it will use our `supplementaryViewProvider`, which is defined below.
+
+```swift
+func supplementaryView(in collectionView: UICollectionView, for kind: String, indexPath: IndexPath) -> UICollectionReusableView? {
+        let elementKind = UICollectionView.ElementKind(rawValue: kind)
+
+        switch elementKind {
+        case .badge:
+            let badgeView: BadgeSupplementaryView = collectionView.dequeueReusableSupplementaryView(of: .badge,
+                                                                                                    for: indexPath, configuredWith: String(Int.random(in: 0..<10)))
+            return badgeView
+
+        default: return nil
+        }
+    }
+```
+
+We check to see the `ElementKind` that is being requested by the `UICollectionView`, before dequeuing the appropriate view based on that kind. The re-use identifier is handled by `UtiliKit`, and so the only information required here is the `ElementKind`. In the case that we receive an `ElementKind` that we do not recognize, we will return nil from our provider.
```

Comment:
> In the case that we receive an `ElementKind` that we do not recognize, we will return nil from our provider.

Should we also mention that `fatalError()` could be an acceptable alternative to returning `nil`? If our supplementary view provider receives an `ElementKind` that it doesn't recognize, that means that we've screwed up our overall setup/implementation (a.k.a. "developer error), right? It could even be good to make that the default behavior, although I can see the argument for leaving ultimate control up to the consumer.

Reply:
I debated putting that the `nil` will normally cause a `UIKit` assertion, so maybe that's a good in between? "You can put whatever you want here, but if UIKit isn't happy it's going to crash"